### PR TITLE
[Filesystem] Add targetIterator for fs mirror

### LINKED ISSUE #14068: [Filesystem] mirror() does not work correctly with custom iterators
The `mirror()` method does not work correctly with custom iterators. This allows a user to easily mirror a directory based on certain criteria (for instance, only `.jpg` files, or files larger than 6MB etc etc).

The third parameters of `mirror()` allows us to specify a custom traversable instance which is used in two occasions: to iterate the files that needs to be copied, and to iterate files that needs to be deleted.

The issue is that when deleting files, it needs to traverse inside the TARGET directory, while when using the iterator for copying files, it will traverser inside the ORIGIN directory.

Using the same iterator is not possible in those cases.

### PR DESCRIPTION
**Pull request by danepowell**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature? yes
| Deprecations? | no
| Tickets       | Fix #14068
| License       | MIT

The existing `$iterator` parameter is useless and will not behave as expected if `$options['delete'] == true`. There needs to be a separate iterator parameter for the target directory.

### CODE REVIEW

**Comment by Nyholm** [comment]
Thank you. Could you add a test for this please? Also make sure Fabbot is happy.

**Comment by danepowell** [comment]
Thanks, will do on Monday

**Comment by danepowell** [comment]
@Nyholm can you help me understand why tests for this pass on my local (`./phpunit src/Symfony/Component/Filesystem/`) but fail in CI? It appears that the Finder component is missing in CI.

**Comment by derrabus** [comment]
> It appears that the Finder component is missing in CI.

https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Filesystem/composer.json

You can only use packages that appear in the component's dependencies.

**Comment by Nyholm** [comment]
@derrabus is correct. You could fix that by adding Finder in the require-dev section. 

Or even better, use the `glob()` to avoid the extra dependency.

**Comment by danepowell** [comment]
Thanks! This is ready for review. I don't believe the remaining failure is related to this PR.

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (Nyholm) [inline src/Symfony/Component/Filesystem/Filesystem.php]: I really hate that PHP have made parameter names part of the BC contract with "named parameters". 

Symfony does not include that in the BC promise, however I dont think we should rename parameters just for fun =)
Will this PR be much worse of if we didnt rename this variable?

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (Nyholm) [inline src/Symfony/Component/Filesystem/Filesystem.php]: Im sorry for asking. But cloud we just do something simliar to this instead for introducing `$targetIterator`
```suggestion
            $deleteIterator = null === $iterator ? null : clone $iterator;
```

**Review by Nyholm — COMMENTED** [review]
Thank you! I have two questions that will help me understand the PR and the issue better.

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (danepowell) [reply src/Symfony/Component/Filesystem/Filesystem.php]: We certainly don't _have_ to rename anything, I just think it would make the distinction between the two iterators more obvious.

Is renaming a parameter likely to break anything downstream? I would have thought not. The goal here was to preserve backwards compatibility (otherwise I would have also changed the parameter order).

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (danepowell) [reply src/Symfony/Component/Filesystem/Filesystem.php]: I'm not sure I understand. It looks like the only change here is to clone `$iterator` instead of assigning it, if it's set. Does that change anything functionally?

I thought iterators all included absolute paths, which means you can't reuse an iterator across multiple parent directories. This is well beyond my comfort zone though, I could be wrong. Of course it would be ideal if an iterator _could_ be relative to a parent directory, then we wouldn't need two separate parameters.

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (derrabus) [inline src/Symfony/Component/Filesystem/Filesystem.php]: The `Filesystem` class is not final. We cannot add new parameters to the signature because that would break code that overrides this method.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (danepowell) [reply src/Symfony/Component/Filesystem/Filesystem.php]: I don't see any way to solve this without an extra parameter (though we certainly don't have to rename the existing parameter). What do you propose as an alternative fix?

Comment (derrabus) [reply src/Symfony/Component/Filesystem/Filesystem.php]: You can leave the parameter commented out and use `func_num_args()` to check if it has been passed and `func_get_arg()` to fetch its value. We can truly add the parameter in Symfony 6 then and remove the `func_num_args()`/`func_get_arg()` quirk again.

But this is a pattern we usually apply in feature releases. We would basically introduce a deprecation this way and that should not happen in a bugfix release. Adding new parameters to public methods sounds like we're doing bit more than just a bugfix here.

Comment (danepowell) [reply src/Symfony/Component/Filesystem/Filesystem.php]: I agree it straddles bug fix and feature. Fundamentally, this is definitely a bug (that the iterator cannot be used as intended or documented), but there's no way to fix it (that I see) except to add a second parameter.

I'll attempt the `func_get_arg()` change as you describe, instead of adding the parameter to the function definition. I think this can still happen as a bug fix, the deprecation can happen as a follow-up issue. Let me know if you disagree.

**Comment by danepowell** [comment]
Okay, I've rewritten this in a way that's backwards-compatible.

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (derrabus) [inline src/Symfony/Component/Filesystem/Filesystem.php]:
```suggestion
            $deleteIterator = 4 < \func_num_args() ? func_get_arg(4) : $iterator;
```

We should catch this parameter even if the caller passes 42 arguments to this method.

File: src/Symfony/Component/Filesystem/composer.json
Comment (derrabus) [inline src/Symfony/Component/Filesystem/composer.json]:
```suggestion
        "symfony/finder": "^4.4|^5.0"
```
4.0 is dead an buried. 😃

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
We need to decide if we want to merge this change as a bugfix. @symfony/mergers I need a second opinion for this.

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php
Comment (derrabus) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php:1249]: Should we also add a test case that operates on two iterators?

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (stof) [reply src/Symfony/Component/Filesystem/Filesystem.php]: if downstream uses named arguments for the call, the renaming is a BC break

**Comment by danepowell** [comment]
I'm happy to take any approach you decide, just let me know how you'd like to proceed.

**Comment by TravisCarden** [comment]
Hey, fancy meeting you here, @danepowell! 😄

This is biting me, too, on v5.3.4. @derrabus, can I do anything to move this forward?

**Comment by derrabus** [comment]
We've revised our policy regarding bugfixes recently, and I suppose we need to schedule this for 5.4 as a feature. Can you please do a rebase?

**Comment by grasmash** [comment]
Will this be merged?

**Comment by grasmash** [comment]
@danepowell looks like it needs a rebase.

**Comment by danepowell** [comment]
I rebased on 6.1 and tests are passing again.

File: src/Symfony/Component/Filesystem/composer.json
Comment (derrabus) [inline src/Symfony/Component/Filesystem/composer.json]:
```suggestion
        "symfony/finder": "^5.4|^6.0"
```

File: src/Symfony/Component/Filesystem/Filesystem.php
Comment (derrabus) [inline src/Symfony/Component/Filesystem/Filesystem.php]:
```suggestion
            $deleteIterator = (4 < \func_num_args() ? func_get_arg(4) : null) ?? $iterator;
```

I the caller explicitly passed `null`, the behavior should be the same as if the parameter was not specified at all.

**Comment by derrabus** [comment]
My question from April is still open: If we add a second iterator to the signature, we should have a test scenario in which both iterators are set.

**Comment by danepowell** [comment]
Thanks, I've addressed all feedback.

Just to confirm, this will be backported to 5.4 after being merged into 6.1?

**Comment by stof** [comment]
No. New features are never backported.

**Comment by danepowell** [comment]
I'd argue this is a bug, not a new feature. The expected behavior if `$options['delete'] == true` is to delete files in the target directory. But the actual behavior is that files are never deleted, because the origin iterator doesn't apply. That's why we jumped through hoops to make this backwards-compatible.

If we're not going to backport this, the code would be greatly simplified by breaking compatibility and getting rid of that ghost parameter.

**Comment by stof** [comment]
@danepowell breaking BC is not allowed in a minor versions. And major versions are only allowed to break BC with a migration path built for it in the previous minor.
Backporting it in a patch version of 5.4 or no does not change the need for BC.

**Comment by danepowell** [comment]
Okay, we won't backport it. Any chance of getting this merged? Anything at all I can do to help?

**Comment by fabpot** [comment]
I'm not comfortable with this PR. I think cloning the iterator as suggested would be much better. Maybe replacing the origin directory with the target directory after cloning if part of some iterators (not sure if that's always possible). But asking the user to pass a second iterator does not seem very user-friendly.

Update: Now that I reread the code, the current iterator argument is "broken" as it does not use the `originDir` anyway. So, I'm :-1: with this PR. I would deprecate the current iterator argument instead.

**Comment by danepowell** [comment]
@fabpot the problem is that iterators contain absolute paths. If the origin and target directories are different, you _must_ have separate iterators for each. Unless you know of some way to hack the iterator to use relative paths? That's beyond my level of comfort, but if you think it's possible and wise I can investigate.

**Comment by danepowell** [comment]
I can no longer reproduce the original issue here, not because it's fixed but just because so much time has elapsed and I'm no longer as familiar with these components. Between that and the obvious resistance to fixing this, I'm going to close it for now.

**Pull request closed without merging**
